### DISCUSSION TITLE
Support for function calling with GPT models

### DISCUSSION THREAD

**Discussion by christopher-czaban** [chhoumann/quickadd#627] [Q&A]
Hello everybody,

First, I am incredibly happy about QuickAdd. Plugins like this make Obsidian so much more powerful than most other PKM tools!

I was wondering whether it is possible to use the function calling ability of the newer GPT models. I couldn't find anything so far in the documentation. It would be great if it is possible or if it, at least, will be possible in the near future because it is just so incredibly powerful.

Thanks,
Chris

**Comment by chhoumann** [comment]
Coming back to this one very late - sorry about that, and glad you're enjoying QuickAdd.

Function calling shipped in QuickAdd 2.14.0. From a user script you can build an agent with tools:

```js
module.exports = async ({ quickAddApi }) => {
  const agent = quickAddApi.ai.agent({
    model: "gpt-5",
    system: "You are a vault librarian. Ground every claim in the user's notes.",
    tools: { ...quickAddApi.ai.tools.vault({ only: ["read_note", "search_notes"] }) },
  });
  const { text } = await agent.generate({ prompt: "What do my notes say about gardening?" });
  return text;
};
```

You can define your own tools with `quickAddApi.ai.tool({ description, inputSchema, execute })` - JSON Schema parameters plus a JS handler, i.e. real function calling. Built-in tool groups (vault, workspace, system) are opt-in, with confirmation for destructive calls, and structured output is supported via `agent.generate({ prompt, schema })`.

Docs: https://quickadd.obsidian.guide/docs/AIAssistant

Closing as resolved.
